src/platform/workspaceChunkSearch/common/rerankerService.ts

```diff
@@ -0,0 +1,113 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { createServiceIdentifier } from '../../../util/common/services';
+import { CancellationToken } from '../../../util/vs/base/common/cancellation';
+import { raceCancellationError } from '../../../util/vs/base/common/async';
+import { FileChunkAndScore } from '../../chunking/common/chunk';
+import { ILogService } from '../../log/common/logService';
+import { IExperimentationService } from '../../telemetry/common/nullExperimentationService';
+
+export const IRerankerService = createServiceIdentifier<IRerankerService>('IRerankerService');
+
+export interface IRerankerService {
+	readonly _serviceBrand: undefined;
+	/**
+	 * Re-rank a list of file chunks for a natural language query.
+	 */
+	rerank(query: string, documents: readonly FileChunkAndScore[], token: CancellationToken): Promise<readonly FileChunkAndScore[]>;
+	/**
+	 * Whether the remote reranker endpoint is available
+	 */
+	readonly isAvailable: boolean;
+}
+
+
+interface RemoteRerankResultEntry {
+	readonly index: number;
+	readonly relevance_score?: number;
+}
+interface RemoteRerankResponse {
+	readonly results?: readonly RemoteRerankResultEntry[];
+}
+
+function buildQueryPrompt(userQuery: string): string {
+	return '<|im_start|>system\nJudge whether the Document meets the requirements based on the Query and the Instruct provided. Note that the answer can only be "yes" or "no".<|im_end|>\n'
+		+ '<|im_start|>user\n'
+		+ '<Instruct>: Given a web search query, retrieve relevant passages that answer the query\n'
+		+ `<Query>: ${userQuery}\n`;
+}
+
+function wrapDocument(text: string): string {
+	return `<Document>: ${text}<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n`;
+}
+
+export class RerankerService implements IRerankerService {
+	declare readonly _serviceBrand: undefined;
+
+	constructor(
+		@ILogService private readonly _logService: ILogService,
+		@IExperimentationService private readonly _expService: IExperimentationService
+	) { }
+
+	private get _endpoint(): string | undefined {
+		return this._expService.getTreatmentVariable<string>('rerankEndpointUrl')?.trim();
+	}
+
+	get isAvailable(): boolean {
+		return !!this._endpoint;
+	}
+
+	async rerank(query: string, documents: readonly FileChunkAndScore[], token: CancellationToken): Promise<readonly FileChunkAndScore[]> {
+		if (!documents.length || !this.isAvailable || !this._endpoint) {
```

Comment:
Should we warn if you call this service and it's not available?

Reply:
maybe we throw?